rfd/xxxx-gcp-server-discovery.md

```diff
@@ -0,0 +1,285 @@
+
+The Discovery Service's service account will require the following permissions:
+
+- `compute.instances.setMetadata`
+- `compute.instances.get`
+- `compute.instances.list`
+
+### GCP join method
+
+In order to register GCP virtual machines, a new `gcp` join method will be
+created. The `gcp` join method will be
+an oidc-based join method like `github`, `circleci`, etc. The token will be fetched from the VM's
+instance metadata at the internal URL
+`http://metadata/computeMetadata/v1/instance/service-accounts/default/identity?audience=AUDIENCE&format=FORMAT&licenses=LICENSES`,
+with the audience claim set to the name of the Teleport cluster. The issuer
+will always be `https://accounts.google.com`. The rest of the registration
+process will be identical to that of the
+[other oidc join methods](https://github.com/gravitational/teleport/blob/master/rfd/0079-oidc-joining.md#auth-server-support).
+
+The joining VM will need a [service account](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances)
+assigned to it to be able to generate id tokens. No permissions on the account
+are needed.
+
+#### Teleport Configuration
+
+The existing provision token type can be extended to support GCP
+authentication, using new GCP-specific fields in the token rules section.
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: example_gcp_token
+spec:
+  roles: [Node, Kube, Db]
+  gcp:
+    allow:
+      # IDs of projects from which nodes can join. At least one required.
+      - project_ids: ['p1', 'p2']
+        # Regions and/or zones from which nodes can join. If empty or omitted,
+        # nodes from any location are allowed.
+        locations: ['l1', 'l2']
+        # Emails associated with service accounts assigned to joining nodes. If
+        # empty or omitted, nodes with any email are allowed.
+        emails: ['e1@example.com', 'e2@example.com']
+        # Names of joining instances. If empty or omitted, nodes with any name
+        # are allowed.
+        instance_names: ['i1', 'i2']
+```
+
+The `google.compute_engine.project_id` and `google.compute_engine.zone` JWT
+claims will be mapped to configuration values.
+
+teleport.yaml on the nodes should be configured so that they will use the GCP
+join token:
+
+```yaml
+teleport:
+  join_params:
+    token_name: 'example_gcp_token'
+    method: gcp
+```
+
+### teleport.yaml generation
+
+The `teleport node configure` subcommand will be used to generate a
+new /etc/teleport.yaml file:
+
+```sh
+teleport node configure
+    --auth-server=auth-server.example.com [auth server that is being connected to]
+    --token="$1" # name of the join token, passed via parameter from run-command
+    --labels="teleport.internal/projectId=${PROJECT_ID},\
+    teleport.internal/zone=${ZONE},\
+    teleport.internal/discovered-node=yes,\
+    teleport.internal/discovered-by=$2,\
+    teleport.internal/origin=cloud" # sourced from instance metadata
```

Comment:
Instead of setting these labels via config, can we implement the approach of getting them via API described in https://github.com/gravitational/teleport/pull/22033 right away?